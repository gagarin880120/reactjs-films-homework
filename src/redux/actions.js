function searchResultsAction(results) {
  return {
      type: 'SEARCH',
      results: results
  };
}

function genresAction(genres) {
  return {
      type: 'GENRES',
      genres: genres
  };
}

export { searchResultsAction, genresAction };
