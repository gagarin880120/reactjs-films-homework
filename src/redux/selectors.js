import getGenresString from '../helpers/helpers';

const searchResultsSelector = (state) => {
  state.searchResults.map((v) => {
    const obj = v;
    obj.genres = getGenresString(v.genre_ids, state.genres);
    return { ...obj };
  });
  return state.searchResults;
};

const modalSelector = (state) => state.isModalOpen;

const trailerSelector = (state) => state.trailerURL;


export { searchResultsSelector, modalSelector, trailerSelector };
