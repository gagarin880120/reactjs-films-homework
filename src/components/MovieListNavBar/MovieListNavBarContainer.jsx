import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import MovieListNavBar from './MovieListNavBar';
import { setViewMode, setCurrentAPIRequest } from '../../redux/actions';
import {
  genresSelector, viewModeSelector, currentAPIRequestSelector,
} from '../../redux/selectors';

export function MovieListNavBarContainer({
  onTrending, onTopRated, onUpcoming, genres, onGenreChange,
  switchViewMode, viewMode, currentAPIRequest,
}) {
  return (
    <MovieListNavBar
      onTrending={onTrending}
      onTopRated={onTopRated}
      onUpcoming={onUpcoming}
      onGenreChange={onGenreChange}
      genres={genres}
      switchViewMode={switchViewMode}
      viewMode={viewMode}
      currentAPIRequest={currentAPIRequest}
    />
  );
}

export const mapStateToProps = (state) => ({
  genres: genresSelector(state),
  viewMode: viewModeSelector(state),
  currentAPIRequest: currentAPIRequestSelector(state),
});

export const mapDispatchToProps = (dispatch) => ({
  onTrending() {
    dispatch(setCurrentAPIRequest('popular'));
  },
  onTopRated() {
    dispatch(setCurrentAPIRequest('top_rated'));
  },
  onUpcoming() {
    dispatch(setCurrentAPIRequest('upcoming'));
  },
  onGenreChange(genreId) {
    dispatch(setCurrentAPIRequest(`genre=${genreId}`));
  },
  switchViewMode(mode) {
    dispatch(setViewMode(mode));
  },
});

MovieListNavBarContainer.propTypes = {
  onTrending: PropTypes.func,
  onTopRated: PropTypes.func,
  onUpcoming: PropTypes.func,
  onGenreChange: PropTypes.func,
  genres: PropTypes.instanceOf(Array),
  switchViewMode: PropTypes.func,
  viewMode: PropTypes.string,
  currentAPIRequest: PropTypes.string,
};

MovieListNavBarContainer.defaultProps = {
  onTrending: null,
  onTopRated: null,
  onUpcoming: null,
  onGenreChange: null,
  genres: [],
  switchViewMode: null,
  viewMode: '',
  currentAPIRequest: '',
};

export default connect(mapStateToProps, mapDispatchToProps)(MovieListNavBarContainer);
