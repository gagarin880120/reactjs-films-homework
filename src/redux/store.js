import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

const initialState = {
  searchResults: [],
  genres: [],
  isModalOpen: false,
  trailerURL: '',
  isTrailerLoaded: false,
};

export function reducer(state = initialState, action) {
  switch (action.type) {
    case 'SEARCH':
      return { ...state, searchResults: action.results };
    case 'GENRES':
      return { ...state, genres: action.genres };
    case 'MODAL':
      return { ...state, isModalOpen: action.isModalOpen };
    case 'TRAILER':
      return { ...state, trailerURL: action.trailerURL };
    case 'IS_TRAILER_LOADED':
      return { ...state, isTrailerLoaded: action.isTrailerLoaded };
    default:
      return state;
  }
}

export const store = createStore(reducer, applyMiddleware(thunk));
