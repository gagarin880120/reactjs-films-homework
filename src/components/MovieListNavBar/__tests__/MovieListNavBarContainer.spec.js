import ShallowRenderer from 'react-test-renderer/shallow';
import React from 'react';
import { MovieListNavBarContainer, mapStateToProps, mapDispatchToProps } from '../MovieListNavBarContainer';
import { setCurrentAPIRequest, setViewMode } from '../../../redux/actions';

jest.mock('../../../redux/actions',() => ({
  setCurrentAPIRequest: jest.fn().mockReturnValue('popular'),
  setViewMode: jest.fn().mockReturnValue('list')
}))

describe('MovieListNavBarContainer', () => {
  describe('component', () => {
    test('should render without crashing', () => {
      const renderer = new ShallowRenderer();
      renderer.render(<MovieListNavBarContainer />);
      const result = renderer.getRenderOutput();
      expect(result).toMatchSnapshot();
    });
  })

  describe('mapStateToProps', () => {
    test('should return the right value', () => {

      const initialState = {
        genres: [{id: 28, name: 'Action'}],
        viewMode: 'list',
        currentAPIRequest: 'popular',
      };

      expect(mapStateToProps(initialState).genres).toEqual([{id: 28, name: 'Action'}]);
      expect(mapStateToProps(initialState).viewMode).toEqual('list');
      expect(mapStateToProps(initialState).currentAPIRequest).toEqual('popular');
    });
  });

  describe('mapDispatchToProps', () => {
    afterEach(() => {
      jest.clearAllMocks();
    })

    test('onTrending should dispatch setCurrentAPIRequest actions', () => {
      const dispatch = jest.fn();
      const {onTrending} = mapDispatchToProps(dispatch);

      onTrending();

      expect(setCurrentAPIRequest).toHaveBeenCalled();
    });

    test('onTopRated should dispatch setCurrentAPIRequest actions', () => {
      const dispatch = jest.fn();
      const {onTopRated} = mapDispatchToProps(dispatch);

      onTopRated();

      expect(setCurrentAPIRequest).toHaveBeenCalled();
    });

    test('onUpcoming should dispatch setCurrentAPIRequest actions', () => {
      const dispatch = jest.fn();
      const {onUpcoming} = mapDispatchToProps(dispatch);

      onUpcoming();

      expect(setCurrentAPIRequest).toHaveBeenCalled();
    });

    test('onGenreChange should dispatch setCurrentAPIRequest actions', () => {
      const dispatch = jest.fn();
      const {onGenreChange} = mapDispatchToProps(dispatch);

      onGenreChange();

      expect(setCurrentAPIRequest).toHaveBeenCalled();
    });

    test('switchViewMode should dispatch setViewMode actions', () => {
      const dispatch = jest.fn();
      const {switchViewMode} = mapDispatchToProps(dispatch);

      switchViewMode();

      expect(setViewMode).toHaveBeenCalled();
    });
  });
});
