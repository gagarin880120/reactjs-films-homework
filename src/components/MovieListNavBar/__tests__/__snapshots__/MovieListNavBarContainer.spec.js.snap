// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieListNavBarContainer component should render without crashing 1`] = `
<MovieListNavBar
  currentAPIRequest=""
  genres={Array []}
  onGenreChange={null}
  onTopRated={null}
  onTrending={null}
  onUpcoming={null}
  switchViewMode={null}
  viewMode=""
/>
`;
